### ISSUE TITLE
Windows: Could not initialize class net.sf.launch4j.config.ConfigPersister

### ISSUE THREAD

**Issue by lilvadim** [fvarrui/JavaPackager#146]
I'm submitting a…

- [x] bug report
- [ ] feature request
- [ ] other

**Short description of the issue/suggestion:**
When packaging windows app there is an error when generating EXE starts
```
Could not initialize class net.sf.launch4j.config.ConfigPersister
```

**Steps to reproduce the issue/enhancement:**

1. Setup some javapackager task with main class
2. run task
3. see the error

**What is the expected behavior?**
Generated folder with app

**What is the current behavior?**
Error

**Please tell us about your environment:**

- JavaPackager version: 1.6.3
- OS version: Windows 8.1 x64
- JDK version: OpenJDK 17.0.2
- Build tool: 
	- [ ] Maven 
	- [x] Gradle 

**Other information**
It doesn't seems to occur with `winrun4j` selected as `exeGenerationTool` but `exe` generated by `winrun4j` doesn't run (tried with bundled JRE), just an error window which asks to contact developer. 

Also, could you tell me, if I didn't generate installer for windows, will folder with binary file be portable?

**Comment by fvarrui** [comment]
Hi @lilvadim!
Please, could you share your `pom.xml` (or at least, JavaPackager config) and Maven output?

**Comment by lilvadim** [comment]
> Hi @lilvadim! Please, could you share your `pom.xml` (or at least, JavaPackager config) and Maven output?

But I use Gradle

**Comment by lilvadim** [comment]
In `build.gradle`:
```groovy 
buildscript {
    dependencies {
        classpath 'io.github.fvarrui:javapackager:1.6.3'
    }
}

apply plugin: 'io.github.fvarrui.javapackager.plugin'

javapackager {
    mainClass = 'ru.nucodelabs.gem.core.StartGemApplication'
    bundleJre = true
    organizationName = "NucodeLabs"
    name = "GEM"
    version = "0.1"
    vmArgs = ["-Djava.library.path=."]
    administratorRequired = false
    additionalResources = [
            file("lib/${libPrefix}forwardsolver.${libExt}"),
            file("lib/${libPrefix}misfit.${libExt}")
    ]
}

task packageMacApp(type: PackageTask, dependsOn: build) {
    useResourcesAsWorkingDir = true
    generateInstaller = true
    macConfig {
        icnsFile = file("src/main/resources/img/gem.icns")
        generateDmg = true
        generatePkg = false
    }
}

task packageWindowsExe(type: PackageTask, dependsOn: build) {
    generateInstaller = true
    winConfig {
        generateSetup = true
        generateMsi = false
        icoFile = file("src/main/resources/img/gem.ico")
        setupMode = SetupMode.askTheUser
        removeOldLibs = true
        disableDirPage = false
        disableFinishedPage = false
        disableWelcomePage = false
        createDesktopIconTask = false
    }
}
```
and output is 
```
> Task :packageWindowsExe
        Using packager io.github.fvarrui.javapackager.packagers.WindowsPackager
        Creating app ...
            Initializing packager ...
                PackagerSettings [outputDirectory=C:\GEM\GEM\build, licenseFile=null, iconFile=null, generateInstaller=true, mainClass=ru.nucodelabs.gem.core.StartGemApplication, name=GEM, displayName=GEM, version=0.1, description=GEM, url=null, administratorRequired=false, organizationName=NucodeLabs, organizationUrl=, organizationEmail=, bundleJre=true, customizedJre=true, jrePath=null, jdkPath=C:\Users\Michael\.jdks\openjdk-17.0.2, additionalResources=[C:\GEM\GEM\lib\forwardsolver.dll, C:\GEM\GEM\lib\misfit.dll], modules=[], additionalModules=[], platform=windows, envPath=null, vmArgs=[-Djava.library.path=.], runnableJar=null, copyDependencies=true, jreDirectoryName=jre, winConfig=WindowsConfig [icoFile=C:\GEM\GEM\src\main\resources\img\gem.ico, headerType=gui, companyName=NucodeLabs, copyright=NucodeLabs, fileDescription=GEM, fileVersion=1.0.0.0, internalName=GEM, language=null, originalFilename=GEM.exe, productName=GEM, productVersion=1.0.0.0, trademarks=null, txtFileVersion=0.1, txtProductVersion=0.1, disableDirPage=false, disableProgramGroupPage=true, disableFinishedPage=false, disableRunAfterInstall=true, disableWelcomePage=false, createDesktopIconTask=false, generateSetup=true, generateMsi=false, generateMsm=false, msiUpgradeCode=e3213dda-d6f3-4191-804d-414c495a918d, wrapJar=true, setupLanguages={english=compiler:Default.isl, spanish=compiler:Languages\Spanish.isl}, setupMode=askTheUser, signing=null, registry=Registry [entries=[]], removeOldLibs=true, exeCreationTool=launch4j], linuxConfig=null, macConfig=null, createTarball=false, createZipball=false, extra={}, useResourcesAsWorkingDir=true, assetsDir=C:\GEM\GEM\assets, classpath=null, jreMinVersion=null, manifest=Manifest [additionalEntries={}, sections=[]], additionalModulePaths=[], fileAssociations=null, packagingJdk=C:\Users\Michael\.jdks\openjdk-17.0.2, scripts=Scripts [bootstrap=null, preInstall=null, postInstall=null]]
            Packager initialized!
            
            Creating app structure ...
                Removing folder [C:\GEM\GEM\build\GEM]
                Old app folder removed C:\GEM\GEM\build\GEM
                App folder created: C:\GEM\GEM\build\GEM
                Assets folder created: C:\GEM\GEM\build\assets
            App structure created!
            
            Resolving resources ...
                No license file specified
                Icon file resolved: C:\GEM\GEM\src\main\resources\img\gem.ico
                Effective additional resources [C:\GEM\GEM\lib\forwardsolver.dll, C:\GEM\GEM\lib\misfit.dll, C:\GEM\GEM\src\main\resources\img\gem.ico]
            Resources resolved!
            
            Copying additional resources
                Copying file [C:\GEM\GEM\lib\forwardsolver.dll] to folder [C:\GEM\GEM\build\GEM]
                Copying file [C:\GEM\GEM\lib\misfit.dll] to folder [C:\GEM\GEM\build\GEM]
                Copying file [C:\GEM\GEM\src\main\resources\img\gem.ico] to folder [C:\GEM\GEM\build\GEM]
            All additional resources copied!
            
            Copying all dependencies ...
            Dependencies copied to C:\GEM\GEM\build\GEM\libs!
            
            Creating runnable JAR...
            Runnable jar created in C:\GEM\GEM\build\GEM-0.1-runnable.jar!
            
            Bundling JRE ... with C:\Users\Michael\.jdks\openjdk-17.0.2
                Creating customized JRE ...
                Getting required modules ... 
                    Executing command: cmd.exe /s /c "C:\Users\Michael\.jdks\openjdk-17.0.2\bin\jdeps -q --multi-release 17 --ignore-missing-deps --print-module-deps C:\GEM\GEM\build\GEM\libs\*.jar C:\GEM\GEM\build\GEM-0.1-runnable.jar"
                    java.base,java.desktop,java.scripting,java.sql,jdk.jfr,jdk.unsupported
                Required modules found: [java.base, java.desktop, java.scripting, java.sql, jdk.jfr, jdk.unsupported]
                
                Creating JRE with next modules included: java.base,java.desktop,java.scripting,java.sql,jdk.jfr,jdk.unsupported
                Using C:\Users\Michael\.jdks\openjdk-17.0.2\jmods modules directory
                Executing command: cmd.exe /s /c "C:\Users\Michael\.jdks\openjdk-17.0.2\bin\jlink --module-path C:\Users\Michael\.jdks\openjdk-17.0.2\jmods --add-modules java.base,java.desktop,java.scripting,java.sql,jdk.jfr,jdk.unsupported --output C:\GEM\GEM\build\GEM\jre --no-header-files --no-man-pages --strip-debug --compress=2"
                Removing folder [C:\GEM\GEM\build\GEM\jre\legal]
            JRE bundled in C:\GEM\GEM\build\GEM\jre!
            
            Creating windows EXE ...
                Exe manifest file generated in C:\GEM\GEM\build\assets\GEM.exe.manifest!
                Copying file [C:\GEM\GEM\build\assets\GEM.exe.manifest] to file [C:\GEM\GEM\build\launch4j\app.exe.manifest]
                Copying file [C:\GEM\GEM\src\main\resources\img\gem.ico] to file [C:\GEM\GEM\build\launch4j\app.ico]
                Copying file [C:\GEM\GEM\build\GEM-0.1-runnable.jar] to file [C:\GEM\GEM\build\launch4j\app.jar]

> Task :packageWindowsExe FAILED

Execution failed for task ':packageWindowsExe'.
> Could not initialize class net.sf.launch4j.config.ConfigPersister

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
```

**Comment by lilvadim** [comment]
I think I could create separate issue for `winrun4j`, because as I mentioned, executable doesn't start with/without installation. Should I?

**Comment by fvarrui** [comment]
> I think I could create separate issue for `winrun4j`, because as I mentioned, executable doesn't start with/without installation. Should I?

Yes, please! As they are different issues, it's better to deal with them in separately. Thanks!

**Comment by lilvadim** [comment]
I forgot to mention, I use **JavaFX 17.0.2 (OpenJFX)** in my project. Maybe it's important for that issue.

**Comment by lilvadim** [comment]
> Which version of Gradle are you using? 7.3 as it's the version compatible with Java 17?

Yes, 7.3

**Comment by fvarrui** [comment]
Ok, it seems to be an incompatibility with launch4j Gradle plugin (more specifically, with one of its dependencies) and Java 17. I've upgraded it from 2.4.7 to 2.5.1, and the problem seems to be fixed. I hope this upgrade doesn't introduce other issues.

Changes released to [issue-147 branch](https://github.com/fvarrui/JavaPackager/tree/issue-146). Now you can build and install the plugin manually in your local repo, as it's a SNAPSHOT version (v1.6.4-SNAPSHOT): 

```bash
git clone https://github.com/fvarrui/JavaPackager.git --branch issue-146
cd JavaPackager
gradlew publishToMavenLocal
```

As JavaPackager uses Gradle v6.8.3, it probably doesn't work with Java 17 ... I'm using Java 11 for this.

Please, try it when you can and give me some feedback. Thanks!

**Comment by fvarrui** [comment]
[Issue #119 was related to this](https://github.com/fvarrui/JavaPackager/issues/119) (Update to gradle-launch4j breaks some existing builds).

**Comment by fvarrui** [comment]
JavaPackager 1.6.4 released to Maven Central.

**Comment by lilvadim** [comment]
That's fixed but version 1.6.4 building unrunnable app for MacOS (M1 aarch64), it's crashing. 1.6.3 works fine tho.
I'll create issue tomorrow I think.

**Comment by fvarrui** [comment]
> That issue was fixed but version 1.6.4 is building unrunnable app for MacOS (M1 aarch64), it's crashing. 1.6.3 works fine tho. I'll create issue tomorrow I think.

I've been thinking about this new issue and it could be caused when codesigning, as now hardened is enabled by default in order to notarize the app.

Could you open a new issue and explain what is happening, please?
